students/rmart300/lesson10/donor_collection.py

```diff
@@ -0,0 +1,128 @@
+        print(title)
+
+        #sort the dictionary by descending order of the sum of values
+        sorted_list = sorted(self.donor_list, key=lambda d: d.donation_total, reverse=True)
+        for donor in sorted_list:
+            data_row = "{0:20}  ${1:>15}   {2:>10}   ${3:>15}".format(str(donor),
+                str(donor.donation_total), str(donor.donation_count), str(donor.donation_average))
+            print(data_row)
+
+    def write_letters(self):
+        """ write letters to every donor in dict """
+
+        #clear files from directory before creating new ones
+        for f in os.listdir(DonorCollection.letter_directory):
+            if '.txt' in f:
+                os.remove(DonorCollection.letter_directory + f)
+
+        for donor in self.donor_list:
+            donation_num = 1
+            for donation_amt in donor.amount_list:
+                message = f"Dear {donor.first_name} {donor.last_name},\n\n    "
+                message += f"Thank you for your very kind donation of ${donation_amt}.\n\n    "
+                message += f"It will be put to very good use.\n\n    Sincerely,\n"
+                message += f"        -The Team"
+                with open(f"{DonorCollection.letter_directory}{donor.first_name}_{donor.last_name}_{donation_num}.txt",'w') as f:
+                    f.write(message)
+                donation_num += 1
+
+    @staticmethod
+    def challenge(x,factor):
+        #print(f"{x}*{factor}={x*factor}")
+        return x*factor
+    
+    def greater_than_min_value(self, x):
+        return x >= self.min_value
+
+    def less_than_max_value(self, x):
+        return x <= self.max_value
+
+    def donation_challenge(self):
+        question = "By what factor would you like to increase the donations? "
+        my_factor = int(input(question))
+        question = "What is the minimum allowed amount? "
+        self.min_value = int(input(question)) if question is not None else 0
+        question = "What is the maximum allowed amount? "
+        self.max_value = int(input(question)) if question is not None else 0
+        
+
+        for donor in self.donor_list:
+            amount_list = donor.amount_list[:]
+            #print(donor,"before filters",amount_list)
+            amount_list = list(filter(self.greater_than_min_value, amount_list))
+            amount_list = list(filter(self.less_than_max_value, amount_list))
+            #print(donor,"after filters",amount_list)
+
+            factor_list = [my_factor]*len(amount_list)
+            challenge_list = list(map(DonorCollection.challenge, amount_list, factor_list))
+            amount_list = challenge_list[:]
+            #print(amount_list)
+            challenge_sum = reduce(lambda x,y: x+y, amount_list)
+            print(f"If you limited the donations between {self.min_value} and {self.max_value}")
+            print(f"and increased by a factor of {my_factor}, then the total donation for {donor} would be {challenge_sum}!")
+
+        
```

Comment:
note that in lesson9 where mailroom OO was introduced, the directions hinted to ensure no user interaction code is part of the data handling class.
the idea is that `DonorCollection` only deals with handling donor collection, and if for example we wanted to create mailroom UI instead of mailroom CLI, you could use this class without much change.